client/comms/wsconn.go

```diff
@@ -0,0 +1,377 @@
+	conn.connectedMtx.Lock()
+	conn.connected = connected
+	conn.connectedMtx.Unlock()
+}
+
+// logRoute logs a request keyed by its id.
+func (conn *WsConn) logRequest(id uint64, req *msgjson.Message) {
+	conn.reqMtx.Lock()
+	conn.req[id] = req
+	conn.reqMtx.Unlock()
+}
+
+// FetchRequest fetches the request associated with the id. The returned
+// request is removed from the cache.
+func (conn *WsConn) FetchRequest(id uint64) (*msgjson.Message, error) {
+	conn.reqMtx.Lock()
+	defer conn.reqMtx.Unlock()
+	req := conn.req[id]
+	if req == nil {
+		return nil, fmt.Errorf("no request found for id %d", id)
+	}
+
+	delete(conn.req, id)
+
+	return req, nil
+}
+
+// NextID returns the next request id.
+func (conn *WsConn) NextID() uint64 {
+	return atomic.AddUint64(&conn.rID, 1)
+}
+
+// connect attempts to establish a websocket connection.
+func (conn *WsConn) connect() error {
+	scheme := "ws"
+
+	if conn.tlsCfg != nil {
+		scheme = "wss"
+	}
+
+	url := url.URL{
+		Scheme: scheme,
+		Host:   conn.cfg.Host,
+		Path:   conn.cfg.Path,
+	}
+
+	dialer := &websocket.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: 10 * time.Second,
+	}
+
+	if conn.tlsCfg != nil {
+		dialer.TLSClientConfig = conn.tlsCfg
+	}
+
+	var err error
+	conn.ws, _, err = dialer.Dial(url.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	conn.setConnected(true)
+	conn.ws.SetPingHandler(func(string) error {
+		now := time.Now()
+		err := conn.ws.SetReadDeadline(now.Add(conn.cfg.PingWait))
+		if err != nil {
+			log.Errorf("unable to set read deadline: %v", err)
+			return err
+		}
+
+		// Respond with a pong.
+		return conn.ws.WriteControl(websocket.PongMessage, []byte{},
+			now.Add(conn.cfg.WriteWait))
+	})
+
+	return nil
+}
+
+// keepAlive maintains an active websocket connection by reconnecting when
+// the established connection is broken. This should be run as a goroutine.
+func (conn *WsConn) keepAlive() {
+	ticker := time.NewTicker(time.Second)
+
+	for {
+		select {
+		case <-ticker.C:
+			if !conn.isConnected() {
+				err := conn.connect()
+				if err != nil {
+					log.Errorf("unable to establish a connection: %v", err)
+				}
+
+				// Run the reconnect sync after reestablishing a connection.
+				if conn.cfg.ReconnectSync != nil {
+					conn.cfg.ReconnectSync()
+				}
+			}
+
+		case <-conn.cfg.Ctx.Done():
+			now := time.Now()
+			closeMsg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")
+			conn.writeMtx.Lock()
+			err := conn.ws.WriteControl(websocket.CloseMessage, closeMsg,
+				now.Add(conn.cfg.WriteWait))
+			conn.writeMtx.Unlock()
+			if err != nil {
+				log.Errorf("unable to close websocket connection: %v", err)
+			}
+			ticker.Stop()
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	msg := new(msgjson.Message)
+
+	for {
+		err := conn.ws.ReadJSON(msg)
```

Comment:
Reusing the same struct for marshaling is dangerous, especially when combined with `omitempty` fields. 

https://play.golang.org/p/uGeMieUD65C

Create a new `msgjson.Message` for each loop.